### ISSUE TITLE
Large Audits backlog starves all page-facing tool calls

### ISSUE THREAD

**Issue by zjh7890** [ChromeDevTools/chrome-devtools-mcp#2556]
### Description of the bug

When `chrome-devtools-mcp` connects to a long-lived Chrome session whose page targets have accumulated a large DevTools Audits backlog, page-facing MCP tools consistently time out even though the browser-level CDP endpoint and every individual page target remain responsive.

In the affected session, the MCP process repeatedly writes:

```text
No handler registered for issue code PerformanceIssue
```

This looks like a regression or broader consequence of #1850 / #1852: the warning is present again in v1.7.0, and a sufficiently large initial `Audits.issueAdded` replay appears to starve normal tool requests. It is also related to the eager all-tab work discussed in #1921, but reproduces with only 32 page targets rather than hundreds or thousands.

The following page-facing tools were tested and all timed out in the affected MCP session:

- `list_pages`
- `evaluate_script`
- `list_network_requests`

An MCP-level ping can still succeed while these tools are unavailable.

#### Diagnostic evidence

- Chrome exposed 32 normal page targets.
- Direct browser-level CDP `Target.getTargets`, `Target.attachToTarget`, and `Runtime.evaluate` succeeded for every page; each page probe completed in under one second.
- Enabling the `Audits` domain directly replayed 17,357 `Audits.issueAdded` events, mostly `CookieIssue`, `GenericIssue`, and `PerformanceIssue`.
- After a five-second warm-up, only five new audit events arrived during the next ten seconds. This indicates that the large burst was primarily retained history rather than continuous page traffic.
- A public trigger was verified separately on `https://example.com`: 500 synchronous `document.cookie` reads produced exactly 500 `PerformanceIssue/DocumentCookie` events.

No page target was crashed, discarded, or unresponsive during the direct-CDP control test.

### Reproduction

Public-only reproduction shape:

1. Start Chrome with remote debugging enabled.
2. Open multiple tabs at `https://example.com`.
3. In each tab, run the following in DevTools Console to accumulate `PerformanceIssue/DocumentCookie` entries:

   ```js
   for (let i = 0; i < 1000; i++) {
     void document.cookie;
   }
   ```

   Repeat across enough tabs to build an Audits backlog of roughly 15,000-20,000 events. The backlog size can be confirmed through a separate raw-CDP session by enabling `Audits` and counting `Audits.issueAdded` events.

4. Connect `chrome-devtools-mcp` to that Chrome instance using either `--autoConnect` or a browser WebSocket endpoint.
5. Call `list_pages`.
6. Call a page-specific tool such as `evaluate_script` with a trivial function.

Observed behavior:

- Tool calls do not return before the client's 60-second timeout.
- stderr repeatedly reports the unhandled `PerformanceIssue` warning.
- Direct CDP calls to the same browser and pages continue to complete normally.
- Restarting only the MCP process does not remove the retained Audits backlog, so reconnecting replays it again.

### Expectation

A large retained Audits backlog, including unknown or unsupported issue codes, should not block unrelated MCP tool calls.

Possible safeguards include bounding/deduplicating initial issue replay, dropping unsupported issue codes before mapping/logging, and ensuring Audits processing cannot starve tool requests.

### MCP configuration

```json
{
  "mcpServers": {
    "chrome-devtools": {
      "command": "npx",
      "args": [
        "-y",
        "chrome-devtools-mcp@1.7.0",
        "--autoConnect"
      ]
    }
  }
}
```

The same behavior was observed with an explicitly supplied loopback `--wsEndpoint`; the endpoint identifier is intentionally omitted.

### Chrome DevTools MCP version

1.7.0

### Chrome version

151.0.7922.76

### Coding agent version

N/A. The failure was observed at the MCP tool-call boundary and compared against raw CDP calls.

### Model version

N/A

### Chat log

N/A

### Node version

22.22.2

### Operating system

macOS 26.5.1

### Extra checklist

- [ ] I want to provide a PR to fix this bug

**Comment by OrKoN** [comment]
@nroscino could you please try to create a repro as a standalone script?

**Comment by nroscino** [comment]
sure @OrKoN! Will try to repro later during the afternoon!

**Comment by cylixlee** [comment]
I've encoutered with this issue too.

**Comment by nroscino** [comment]
Hello @cylixlee @zjh7890! I have been trying to reproduce this but cannot really make it timeout.

This is the script I am using to generate the pages with issues:

```javascript
import puppeteer from "puppeteer";

async function run() {
  const browser = await puppeteer.launch({
    headless: false,
    args: ["--remote-debugging-port=9222"],
  });

  const wsEndpoint = browser.wsEndpoint();
  console.log(`\nBrowser WebSocket Endpoint: ${wsEndpoint}`);

  const NUM_TABS = 400;
  const EVENTS_PER_TAB = 1000;

  const pages = [];

  while (pages.length < NUM_TABS) {
    const page = await browser.newPage();
    await page.goto("https://example.com");
    pages.push(page);
  }

  let actualIssueCount = 0;
  const clients = [];

  console.log("Generating issues...");

  for (let i = 0; i < pages.length; i++) {
    const page = pages[i];

    const client = await page.createCDPSession();
    clients.push(client);

    client.on("Audits.issueAdded", () => {
      actualIssueCount++;
    });
    await client.send("Audits.enable");

    await page.evaluate((limit) => {
      for (let j = 0; j < limit; j++) {
        document.cookie = `test${j}=1; SameSite=None`;
      }
    }, EVENTS_PER_TAB);

    await new Promise((r) => setTimeout(r, 50));
  }

  for (const client of clients) {
    await client.detach();
  }

  console.log(`Setup complete. Verified ${actualIssueCount} Audits events were generated in the browser.`);
  console.log(
    `DEBUG=* npx @modelcontextprotocol/inspector npx -y chrome-devtools-mcp@latest --wsEndpoint=${wsEndpoint}`,
  );

  await new Promise(() => {});
}

run().catch(console.error);
```

@OrKoN can you also try to repro?

**Comment by OrKoN** [comment]
@nroscino it does not timeout for me but I also do not get any issues in list_console_messages do you get any?

**Comment by cylixlee** [comment]
I was using DeepSeek Harness (WebUI) and I told it to open YouTube using the Chrome DevTools MCP. Chrome did open the website but the stderr is full of `No handler registered for issue code PerformanceIssue`.

Here's my reproduce: https://github.com/cylixlee/chrome-devtools-mcp-reproduce

**Comment by OrKoN** [comment]
I believe any messages in stderr should not have any effect on functionality. Stderr can be used for logging according to the MCP spec. Also, https://github.com/ChromeDevTools/chrome-devtools-mcp/pull/2535 might have fixed it.

**Comment by nroscino** [comment]
> [@nroscino](https://github.com/nroscino) it does not timeout for me but I also do not get any issues in list_console_messages do you get any?

no I also don't get them even if I explicitly log something, but from the console tab in DT I can see them

**Comment by giova23** [comment]
I can confirm the `No handler registered for issue code PerformanceIssue` stderr spam persists on **v1.8.0** (fresh npm tarball; Chrome 151.0.7922.169, Node 22.22.1). The guard from #2535 covers only one of the two consumers of `Audits.issueAdded`:

- **Guarded (silent):** the collector's own subscription — `PageEventSubscriber.#onIssueAdded` (`build/src/collectors/PageCollector.js:167`) bails out via `if (!DevTools.isIssueCodeSupported(inspectorIssue.code)) return;`. Since `isIssueCodeSupported(code)` is `issueCodeHandlers.has(code)` and `PerformanceIssue` has no handler entry, this path is correctly silent in 1.8.0.
- **Not guarded (the spam):** the DevTools-frontend-internal pipeline inside the vendored bundle (`build/src/third_party/index.js`): `Audits.issueAdded` → `IssuesModel.issueAdded()` (~line 125208) → `IssueAdded` dispatch → `IssuesManager.#onIssueAddedEvent` (~line 164636) → `createIssuesFromProtocolIssue` (~line 164528), which does `console.warn(`No handler registered for issue code ${code}`)` whenever the code has no handler. #2535 does not touch this path, so every `PerformanceIssue` event (e.g. `DocumentCookie` from heavy `document.cookie` access) still warns. Long-lived Chrome sessions amplify this because `Audits.enable` replays the accumulated issue backlog on each (re)connect.

Since that `console.warn` is the only occurrence of the string in the whole package and the collector path is provably guarded, in v1.8.0 the user-visible stderr spam can only come from this internal pipeline.

Minimal function-level repro (no browser needed, v1.8.0 tarball):

```js
import { DevTools } from '.../node_modules/chrome-devtools-mcp/build/src/third_party/index.js';
DevTools.createIssuesFromProtocolIssue(null, {
  code: 'PerformanceIssue',
  details: { performanceIssueDetails: { performanceIssueType: 'DocumentCookie' } },
});
// → "No handler registered for issue code PerformanceIssue" on stderr, returns []
// (also: DevTools.isIssueCodeSupported('PerformanceIssue') === false)
```

Two candidate fixes:
1. Register a no-op handler for `PerformanceIssue` in `issueCodeHandlers` (returning `[]`) — silences every consumer path at once and makes `isIssueCodeSupported('PerformanceIssue')` true; or
2. Stop `createIssuesFromProtocolIssue` from warning for known-but-unmapped codes (e.g. demote to a debug log, or skip `PerformanceIssue` specifically).

For now I'm running a local one-line patch on the vendored bundle (wrap the `console.warn` in `if (String(inspectorIssue.code) !== "PerformanceIssue")`) until an upstream fix ships. Happy to open a PR if that would help.
